vllm/v1/worker/gpu_model_runner.py

```diff
@@ -3132,6 +3154,29 @@ def sample_tokens(
         with record_function_or_nullcontext("gpu_model_runner: sample"):
             sampler_output = self._sample(logits, spec_decode_metadata)
 
+        # Mask out invalid spec tokens for async scheduling + structured outputs.
+        if (
+            grammar_output is not None
+            and grammar_output.num_invalid_tokens_per_req is not None
+            and self.use_async_scheduling
+        ):
+            num_invalid_tokens_per_req = grammar_output.num_invalid_tokens_per_req
+            num_reqs, num_sampled_toks = sampler_output.sampled_token_ids.shape
+            num_invalid_spec_tokens_cpu = torch.zeros((num_reqs,), dtype=torch.int32)
+            for req_id, num_invalid_toks in zip(
+                grammar_output.structured_output_request_ids, num_invalid_tokens_per_req
+            ):
+                req_index = self.input_batch.req_id_to_index[req_id]
+                num_invalid_spec_tokens_cpu[req_index] = num_invalid_toks
+            col_indices = torch.arange(num_sampled_toks, dtype=torch.int32)
+            mask_start_indices = num_sampled_toks - num_invalid_spec_tokens_cpu
+            mask = col_indices.unsqueeze(0) >= mask_start_indices.unsqueeze(1)
+            self.invalid_spec_tokens_mask.cpu[:num_reqs, :].copy_(mask)
```

Comment:
Should be faster to do this in numpy
```suggestion
            num_invalid = np.zeros((num_reqs,), dtype=np.int32)
            for req_id, num_invalid_toks in zip(
                grammar_output.structured_output_request_ids, num_invalid_tokens_per_req
            ):
                req_index = self.input_batch.req_id_to_index[req_id]
                num_invalid[req_index] = num_invalid_toks
            mask_start_indices = np.subtract(num_sampled_toks, num_invalid, out=num_invalid)
            np.greater_equal(
                np.arange(num_sampled_toks, dtype=np.int32)[None, :],
                mask_start_indices[:, None],
                out=self.invalid_spec_tokens_mask.np[:num_reqs, :]
            )
```